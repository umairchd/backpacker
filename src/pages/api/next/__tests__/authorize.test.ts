/**
 * @jest-environment node
 */

import path from "node:path";

import handler from "../authorize.page";

import { createMocks } from "node-mocks-http";

const TEST_NAME = "authorize";

import nock from "nock";
import { find } from "lodash";

describe(TEST_NAME, () => {
  let nockBack: nock.Back;

  beforeAll(async () => {
    nockBack = nock.back;
    nockBack.setMode(nockBack.currentMode);
    nockBack.fixtures = path.resolve(
      __dirname,
      `./__nock-fixtures__/${TEST_NAME}`
    );
  });

  beforeEach(() => {
    nock.enableNetConnect(/(localhost|127\.0\.0\.1)/);
  });

  afterEach(() => {
    nock.restore();
    nock.cleanAll();
    jest.resetAllMocks();
  });

  it("happy case", async () => {
    Date.now = jest.fn(() => 1678967889713); // this timestamp is captured to pair with the fixtures, it will need to be updated if the fixture needs to be updated

    const { nockDone } = await nockBack("happy_case.json");

    const { req, res } = createMocks({
      method: "POST",
      body: JSON.stringify({
        deviceId: "random random id",
      }) as any,
    });

    await handler(req as any, res as any);

    expect(res._getStatusCode()).toBe(200);
    expect(res._getData()).toMatchSnapshot();

    nockDone();
  });

  it("happy case with authorization header", async () => {
    Date.now = jest.fn(() => 1678967889713); // this timestamp is captured to pair with the fixtures, it will need to be updated if the fixture needs to be updated

    const { nockDone } = await nockBack("happy_case_with_auth_header.json");

    const happyCaseFixture = require("./__nock-fixtures__/authorize/happy_case.json");
    const happyCaseFixtureGuestTokenResponse = find(happyCaseFixture, {
      scope: "https://api-stage.travelloapp.com:443",
      method: "POST",
      path: "/2/users/guest",
    })?.response;

    const { req, res } = createMocks({
      method: "POST",
      body: JSON.stringify({
        deviceId: "random random id",
      }) as any,
      headers: {
        // this access token is captured to pair with the fixtures, it will need to be updated if the fixture needs to be updated
        authorization: `Bearer ${happyCaseFixtureGuestTokenResponse.authorizationToken}`,
      },
    });

    await handler(req as any, res as any);

    expect(res._getStatusCode()).toBe(200);
    expect(res._getData()).toMatchSnapshot();

    nockDone();
  });

  it("fail at GET", async () => {
    const { nockDone } = await nockBack("fail_case.json");

    const { req, res } = createMocks({
      method: "GET",
      query: {
        deviceId: "random random id",
      },
    });

    await handler(req as any, res as any);

    expect(res._getStatusCode()).toBe(400);
    expect(res._getData()).toMatchSnapshot();

    nockDone();
  });
});
