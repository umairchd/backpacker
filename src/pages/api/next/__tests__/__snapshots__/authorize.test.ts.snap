// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorize fail at GET 1`] = `"You have you ask me nicely."`;

exports[`authorize happy case 1`] = `
{
  "debug": {
    "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJFT3BmRU5vUnplZFhpXzBkS01LMHdQMzFsd29TSWdjdThQZUZiZ2dNanFnIn0.eyJleHAiOjE2ODg3MjE0NzksImlhdCI6MTY4ODcyMDU3OSwianRpIjoiNTJmMWQyYTEtMmI4Mi00ZDEzLTljYjYtZGFjZWYzZmJiMzNjIiwiaXNzIjoiaHR0cHM6Ly9zdGFnZS50cmF2ZWxsb2FwcC5jb20vYXV0aC9yZWFsbXMvdHJhdmVsbG8iLCJhdWQiOlsidHJhdmVsbG8iLCJiYWNrcGFja2VyZGVhbHMtbmV4dCJdLCJzdWIiOiI4YTI0YWUzZS02NjI5LTRiYmUtYTgyOS00NDgyN2U5ZDBiNDciLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJiYWNrcGFja2VyZGVhbHMtbmV4dCIsInNlc3Npb25fc3RhdGUiOiJjY2NhZDc3Mi1kNzFmLTRkMmQtYTgwNi1hMjcwZjFiZTBkNzciLCJhY3IiOiIxIiwic2NvcGUiOiJlbWFpbCBwcm9maWxlIHRyYXZlbGxvIiwic2lkIjoiY2NjYWQ3NzItZDcxZi00ZDJkLWE4MDYtYTI3MGYxYmUwZDc3IiwiUk9MRSI6IkdVRVNUIiwiQVBQIjoiVFJBVkVMTE8iLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImdlbmRlciI6IiIsInJvbGVzIjpbImRlZmF1bHQtcm9sZXMtdHJhdmVsbG8iLCJST0xFX1VTRVIiLCJvZmZsaW5lX2FjY2VzcyIsInVtYV9hdXRob3JpemF0aW9uIl0sInByZWZlcnJlZF91c2VybmFtZSI6InJhbmRvbS1yYW5kb20taWRAZ3Vlc3QudHJhdmVsbG9hcHAuY29tIiwiZ2l2ZW5fbmFtZSI6IiIsImZhbWlseV9uYW1lIjoiIiwiZW1haWwiOiJyYW5kb20tcmFuZG9tLWlkQGd1ZXN0LnRyYXZlbGxvYXBwLmNvbSIsImNpZCI6InRyYXZlbGxvIn0.G1O5Ixxn3rPv0W_CZgsMep6eNFl-Fq-dcvqLRw_aE7ga5ueaOxxUANpp-6C7ASbiny0ekk_IShopgSIOzUxovpEPAx2k9W_j4Pa09i6xNT4aXtMZi5HpdjIOl5I5LeeqNYuEVBON5_7LC-H9GGYb6hJtZv0HLaUgNDbTws6k6MzxGhdlW7-BhrdSjA55x5MTl1mbQ8kPtXcJp6S70psTviFKt4wFO18vYDdmm1wcorlpJ-Wb3xdxQkwixKbXwXH4S3qvuC_sq5EH0vcev_l0hNUYfgAh_Mf-SkoNy_JkclhHTP-Vmj1yp_HQ6ASrIEcEzjKp7Wg8FiR6v7u5Q6y7Rw",
    "expires_at": 1678968789,
    "expires_in": 900,
    "refresh_token": "eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI5MDY2YjQ2Yi04YzA0LTQxOWItOGMyOS02N2JjNzIzMmY5ZGYifQ.eyJleHAiOjE2ODg3Mjc3NzksImlhdCI6MTY4ODcyMDU3OSwianRpIjoiYTc0ZDM1YzAtZGEwNC00NGMwLWE1NmUtYWIzOWJlZDk1NGQ2IiwiaXNzIjoiaHR0cHM6Ly9zdGFnZS50cmF2ZWxsb2FwcC5jb20vYXV0aC9yZWFsbXMvdHJhdmVsbG8iLCJhdWQiOiJodHRwczovL3N0YWdlLnRyYXZlbGxvYXBwLmNvbS9hdXRoL3JlYWxtcy90cmF2ZWxsbyIsInN1YiI6IjhhMjRhZTNlLTY2MjktNGJiZS1hODI5LTQ0ODI3ZTlkMGI0NyIsInR5cCI6IlJlZnJlc2giLCJhenAiOiJiYWNrcGFja2VyZGVhbHMtbmV4dCIsInNlc3Npb25fc3RhdGUiOiJjY2NhZDc3Mi1kNzFmLTRkMmQtYTgwNi1hMjcwZjFiZTBkNzciLCJzY29wZSI6ImVtYWlsIHByb2ZpbGUgdHJhdmVsbG8iLCJzaWQiOiJjY2NhZDc3Mi1kNzFmLTRkMmQtYTgwNi1hMjcwZjFiZTBkNzcifQ.R-rd9qMvTvd5TjntxlvSNOKp3YgyDHikUpZsyXgJ_Ps",
  },
  "message": "Authenticated",
  "ok": true,
}
`;

exports[`authorize happy case with authorization header 1`] = `
{
  "debug": {
    "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJFT3BmRU5vUnplZFhpXzBkS01LMHdQMzFsd29TSWdjdThQZUZiZ2dNanFnIn0.eyJleHAiOjE2ODg3MjE0ODAsImlhdCI6MTY4ODcyMDU4MCwianRpIjoiMzkzODBiYTUtNzdlNi00MGE5LWFjNTctNjk3YmUzMGJiM2ZkIiwiaXNzIjoiaHR0cHM6Ly9zdGFnZS50cmF2ZWxsb2FwcC5jb20vYXV0aC9yZWFsbXMvdHJhdmVsbG8iLCJhdWQiOlsidHJhdmVsbG8iLCJiYWNrcGFja2VyZGVhbHMtbmV4dCJdLCJzdWIiOiI4YTI0YWUzZS02NjI5LTRiYmUtYTgyOS00NDgyN2U5ZDBiNDciLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJiYWNrcGFja2VyZGVhbHMtbmV4dCIsInNlc3Npb25fc3RhdGUiOiJjY2NhZDc3Mi1kNzFmLTRkMmQtYTgwNi1hMjcwZjFiZTBkNzciLCJhY3IiOiIxIiwic2NvcGUiOiJlbWFpbCBwcm9maWxlIHRyYXZlbGxvIiwic2lkIjoiY2NjYWQ3NzItZDcxZi00ZDJkLWE4MDYtYTI3MGYxYmUwZDc3IiwiUk9MRSI6IkdVRVNUIiwiQVBQIjoiVFJBVkVMTE8iLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsImdlbmRlciI6IiIsInJvbGVzIjpbImRlZmF1bHQtcm9sZXMtdHJhdmVsbG8iLCJST0xFX1VTRVIiLCJvZmZsaW5lX2FjY2VzcyIsInVtYV9hdXRob3JpemF0aW9uIl0sInByZWZlcnJlZF91c2VybmFtZSI6InJhbmRvbS1yYW5kb20taWRAZ3Vlc3QudHJhdmVsbG9hcHAuY29tIiwiZ2l2ZW5fbmFtZSI6IiIsImZhbWlseV9uYW1lIjoiIiwiZW1haWwiOiJyYW5kb20tcmFuZG9tLWlkQGd1ZXN0LnRyYXZlbGxvYXBwLmNvbSIsImNpZCI6InRyYXZlbGxvIn0.qRdOD4DD0QJK47kYRQWJiDLX30Ugd5DOBCuGE6hzSJAc9u1bYYM-pm1dWpRWsSlQa-IC270J4EZfqkXKE46zpuKRvaF5bpU41ed0z0rhWKuNLTDxbPhJh3VHcPm89-sXuX9ESvKEUvzbXBA_T2PW42Ah9Itu3D26su4prlLRJSrsz23VfHlQyFpZ8ST8rbvsXyyfeaWgjaUENks6DLF8h8ORSVU-BJZwMUtEHiloPBlJz4QVkrZ9kYI7rkzWoH4sZufWiRDmwfTB8QOOzfLujzzx7sPacw8i6pPYypCB-jgwyOqodXKlzthTUSWHVZeP1hmfsE35n9ZTQK1W1E2-tg",
    "expires_at": 1678968789,
    "expires_in": 900,
    "refresh_token": "eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI5MDY2YjQ2Yi04YzA0LTQxOWItOGMyOS02N2JjNzIzMmY5ZGYifQ.eyJleHAiOjE2ODg3Mjc3ODAsImlhdCI6MTY4ODcyMDU4MCwianRpIjoiNGU5MTE3NjgtMjY3YS00NzAwLTliMjItN2Q2ODZiOGZmYmNiIiwiaXNzIjoiaHR0cHM6Ly9zdGFnZS50cmF2ZWxsb2FwcC5jb20vYXV0aC9yZWFsbXMvdHJhdmVsbG8iLCJhdWQiOiJodHRwczovL3N0YWdlLnRyYXZlbGxvYXBwLmNvbS9hdXRoL3JlYWxtcy90cmF2ZWxsbyIsInN1YiI6IjhhMjRhZTNlLTY2MjktNGJiZS1hODI5LTQ0ODI3ZTlkMGI0NyIsInR5cCI6IlJlZnJlc2giLCJhenAiOiJiYWNrcGFja2VyZGVhbHMtbmV4dCIsInNlc3Npb25fc3RhdGUiOiJjY2NhZDc3Mi1kNzFmLTRkMmQtYTgwNi1hMjcwZjFiZTBkNzciLCJzY29wZSI6ImVtYWlsIHByb2ZpbGUgdHJhdmVsbG8iLCJzaWQiOiJjY2NhZDc3Mi1kNzFmLTRkMmQtYTgwNi1hMjcwZjFiZTBkNzcifQ.pwSBfFNSCv8edEzM-0l0p3sW-Ubd29Eh-wmQGbApMKc",
  },
  "message": "Authenticated",
  "ok": true,
}
`;
